### ISSUE TITLE
Multimodal checkpoints with untied lm_head are silently tied to embed_tokens (wrong output head, no warning)

### ISSUE THREAD

**Issue by jiwoon-ahn** [pjordanandrsn/experts4bit-qlora#37]
**Version**: 0.6.3 · **File**: `loader.py`

## Summary

`load_moe_4bit_streaming` builds the text tower from multimodal checkpoints by
stripping the `model.language_model.` prefix and keeping only text-tower keys
(loader.py:123). The top-level `lm_head.weight` has no such prefix, so it is
dropped — and the tie fallback then assigns the embedding matrix as the output
head **unconditionally**:

```python
# loader.py ~:293
# Tie lm_head if the checkpoint relied on weight tying.
if model.lm_head.weight.is_meta:
    model.lm_head.weight = model.model.embed_tokens.weight
```

For tied models (Gemma-4: no `lm_head.weight` on disk) this is correct. For
**untied** multimodal models (`tie_word_embeddings: false` with a real
`lm_head.weight` in the checkpoint index — e.g. Qwen3.6-35B-A3B, vocab 248320,
lm_head [248320, 2048]) every forward computes logits through the wrong matrix.

## Why it's nasty

No error, no warning, plausible-shaped generations. Symptoms we chased for a
day before finding it:

- Base model "generates garbage" identically under nf4 AND int8 (quant-invariant
  — which wrongly indicts quantization fidelity).
- Initial train loss ≈ ln(vocab) (uniform logits), mimicking a chat-template bug.
- LoRA training "converges" — it learns to steer hidden states into the
  embedding matrix — then collapses when served on an inference stack that maps
  lm_head correctly (vLLM): we measured 0.79 → 0.12 accuracy.
- A/B against a tied model (Gemma-4) on "the same code path" passes, because
  the fallback is correct there — the A/B convicts the wrong suspect.

## Suggested fix

Gate the tie on config, load the real head when present, and fail loud otherwise:

```python
if model.lm_head.weight.is_meta:
    if getattr(lm_config, "tie_word_embeddings", True):
        model.lm_head.weight = model.model.embed_tokens.weight
    elif "lm_head.weight" in weight_map:          # top-level, unprefixed key
        _assign(model, "lm_head.weight", get("lm_head.weight"))
    else:
        raise RuntimeError(
            "untied lm_head but no lm_head.weight in checkpoint — refusing to "
            "silently tie to embed_tokens (wrong output head)."
        )
```

(`weight_map` here must be consulted **before** the prefix filter drops
top-level keys — the tensor is in the checkpoint, just not under
`model.language_model.`.)

## Repro sketch

Any `tie_word_embeddings: false` multimodal checkpoint: load via
`load_moe_4bit_streaming`, compare `model.lm_head.weight.data_ptr()` with
`model.model.embed_tokens.weight.data_ptr()` — they are equal (tied) even
though the config says untied and the checkpoint carries a distinct head.
Greedy generations are incoherent vs the same weights under vLLM/transformers.

## Workaround we ship meanwhile

Post-load: re-download/read the `lm_head.weight` shard, assign it, and assert
`data_ptr != embed_tokens` + shape `[vocab, hidden]`.
